webrtc-c/canary/src/Main.cpp

```diff
@@ -0,0 +1,179 @@
+
+    MEMSET(&videoTrack, 0x00, SIZEOF(RtcMediaStreamTrack));
+    MEMSET(&audioTrack, 0x00, SIZEOF(RtcMediaStreamTrack));
+
+    // Declare that we support H264,Profile=42E01F,level-asymmetry-allowed=1,packetization-mode=1 and Opus
+    CHK_STATUS(pPeer->addSupportedCodec(RTC_CODEC_H264_PROFILE_42E01F_LEVEL_ASYMMETRY_ALLOWED_PACKETIZATION_MODE));
+    CHK_STATUS(pPeer->addSupportedCodec(RTC_CODEC_OPUS));
+
+    // Add a SendRecv Transceiver of type video
+    videoTrack.kind = MEDIA_STREAM_TRACK_KIND_VIDEO;
+    videoTrack.codec = RTC_CODEC_H264_PROFILE_42E01F_LEVEL_ASYMMETRY_ALLOWED_PACKETIZATION_MODE;
+    STRCPY(videoTrack.streamId, "myKvsVideoStream");
+    STRCPY(videoTrack.trackId, "myVideoTrack");
+    CHK_STATUS(pPeer->addTransceiver(videoTrack));
+
+    // Add a SendRecv Transceiver of type video
+    audioTrack.kind = MEDIA_STREAM_TRACK_KIND_AUDIO;
+    audioTrack.codec = RTC_CODEC_OPUS;
+    STRCPY(audioTrack.streamId, "myKvsVideoStream");
+    STRCPY(audioTrack.trackId, "myAudioTrack");
+    CHK_STATUS(pPeer->addTransceiver(audioTrack));
+
+CleanUp:
+
+    return retStatus;
+}
+
+VOID sendLocalFrames(Canary::PPeer pPeer, MEDIA_STREAM_TRACK_KIND kind, const std::string& pattern, UINT64 frameCount, UINT32 frameDuration)
+{
+    STATUS retStatus = STATUS_SUCCESS;
+    Frame frame;
+    UINT64 fileIndex = 0, frameSize;
+    CHAR filePath[MAX_PATH_LEN + 1];
+    UINT64 startTime, lastFrameTime, elapsed;
+
+    frame.frameData = NULL;
+    frame.size = 0;
+    frame.presentationTs = 0;
+    startTime = GETTIME();
+    lastFrameTime = startTime;
+
+    while (!terminated.load()) {
+        fileIndex = fileIndex % frameCount + 1;
+        SNPRINTF(filePath, MAX_PATH_LEN, pattern.c_str(), fileIndex);
+
+        CHK_STATUS(readFile(filePath, TRUE, NULL, &frameSize));
+
+        // Re-alloc if needed
+        if (frameSize > frame.size) {
+            frame.frameData = (PBYTE) REALLOC(frame.frameData, frameSize);
+            CHK_ERR(frame.frameData != NULL, STATUS_NOT_ENOUGH_MEMORY, "Failed to realloc media buffer");
+        }
+        frame.size = (UINT32) frameSize;
+
+        CHK_STATUS(readFile(filePath, TRUE, frame.frameData, &frameSize));
+
+        frame.presentationTs += frameDuration;
+
+        pPeer->writeFrame(&frame, kind);
+
+        // Adjust sleep in the case the sleep itself and writeFrame take longer than expected. Since sleep makes sure that the thread
+        // will be paused at least until the given amount, we can assume that there's no too early frame scenario.
+        // Also, it's very unlikely to have a delay greater than SAMPLE_VIDEO_FRAME_DURATION, so the logic assumes that this is always
+        // true for simplicity.
+        elapsed = lastFrameTime - startTime;
+        THREAD_SLEEP(frameDuration - elapsed % frameDuration);
+        lastFrameTime = GETTIME();
+    }
+
+CleanUp:
+
+    SAFE_MEMFREE(frame.frameData);
+
+    auto threadKind = kind == MEDIA_STREAM_TRACK_KIND_VIDEO ? "video" : "audio";
```

Comment:
Curious why we cannot use kind directly.

Reply:
It's just for the sake of readability. By doing this, I can describe my intention through the variable name. Also, I don't need to duplicate the logic for the else statement.